LTI 1.3 tool support for blockstore-based content (aka. "Content Libraries V2") [SE-4514]

jvdm opened the pull request:
## Description

This PR enables Open EdX to act as an LTI 1.3 tool (formerly known as "tool provider") for content managed by Content Libraries and backed up by blockstore. It allows instructors and staff users to embed content within other LMS and platforms, for example, Canvas.

The support for LTI features is limited to a subset of use-cases, as follows:

1. Tool setup and configuration through administration interfaces.
2. Retrieval and rendering of blockstore-based content (XBlocks) in a runtime suitable for LTI resource link launches. That is IFrames within the platform.
3. Send grades associated with content, i.e. assessments, to the platform gradebooks.

Once merged, Open edX will be capable of serving any content managed by Content Libraries to an LTI-supporting platform that was previously configured and authorized.

This PR's technical approach and decision are available in the [ADR for LTI providers in content libraries](https://github.com/edx/edx-platform/pull/27089).

**JIRA:** [SE-4257](https://tasks.opencraft.com/browse/SE-4257)

**Reviewers**:

- [X] @arbrandes
- [ ] @marcotuts 

## Supporting information

[ADR` for LTI providers in content libraries](https://github.com/edx/edx-platform/pull/27089).

## Concerns

There are two. I believe they don't block the current implementation. But, we might need to follow-up later to accommodate future use cases. They are not blockers to those future use cases AFAICT.

First, grade syncing assumes a 1:1 relationship between XBlocks and grade lineitem. Hence, the logic ignores the potential tree relationship of XBlocks and how the tree should be mapped to a single LTI launch. Lack of a deeper understanding of this relationship within content libraries and the whole XBlock use-cases is the main reason for the current approach. Meaning, I understood users of LTI want to expose individual blocks hosting one assessment. This conclusion was the result of my own tests within Content Libraries.

Second, since each assessment can only hold one score, grade syncing is done "synchronously" rather than carried out by celery. Throughout my tests, the grade update event seemed to fail synchronously if the LTI platform was unavailable and it won't block retries. My reasoning is, this is a better user experience than keeping a work log of grades to sync and simplifies the implementation.

## Testing instructions

These are steps to manually test the changes in development environments of openedx as the tool and canvas as the platform.

### 1. Setup devstacks

1. Deploy an instance of [edx-platform devstack](https://github.com/edx/devstack#id8).
2. Deploy an instance of [canvas' development stack](https://github.com/instructure/canvas-lms/wiki/Quick-Start#automated-setup).

   Some notes:

   - Use their automated setup.
   - Make sure to enable ``dory``, which will offer a local DNS resolver and HTTP proxy to route requests to ``canvas.docker``.

3. Deploy an instance of [edx/blockstore](https://github.com/edx/blockstore/#using-with-docker-devstack).
4. Follow the steps at [frontend-app-library-authoring](https://github.com/edx/frontend-app-library-authoring/).

At the end you should have a development instance of Canvas and Open edX, with Content Libraries V2 enabled, in your workstation.

You need to allow routing from edX studio container to the Canvas container.  One straightforward way to achieve that is:

1. Find the host IP in the docker bridge of your workstation:

   ```
   % ip -br addr show docker0
   docker0          UP             1.2.3.4/16 fe80::42:39ff:fec3:140c/64 
   %
   ```

   In the above the IP is `1.2.3.4`.

2. Overwrite `/etc/hosts in the edX studio container:

   ```
   % cd ~/src/edx-devstack
   % docker-compose exec studio bash
   root@studio:/edx/app/edx_ansible/edx_ansible/docker/plays# echo 1.2.3.4 canvas.docker >> /etc/hosts
   root@studio:/edx/app/edx_ansible/edx_ansible/docker/plays# ^D
   %
   ```

This will allow the LTI view to contact the Canvas JWK endpoints to retrieve keys during the LTI launch steps.

3. Finally, disable DJDT in Studio. I personally disabled here: https://github.com/open-craft/edx-platform/blob/master/cms/envs/devstack.py#L99

This is required for the `launch` responses rendering XBlocks to not get tainted by DJDT.

### 2. Setup Canvas

Now, go to `http://canvas.docker/`:

1. Login with the credentials used during Canvas devstack setup.
1. In the left toolbar, click on "Admin" to open the administration panel.
1. Select the the user you created during Canvas devstack setup (do not select "Site Admin").
1. On the menu on the left, click "Developer Keys".
1. Clik on the "+ Developer Key" button, select "+ LTI".
1. On the "method" drop-down, select "Paste JSON".
1. Paste the following JSON:

   ```
    {
        "title": "edX Content Library",
        "scopes": [],
        "extensions": [
            {
                "platform": "canvas.instructure.com",
                "settings": {
                    "platform": "canvas.instructure.com",
                    "placements": [
                        {
                            "placement": "assignment_selection",
                            "message_type": "LtiDeepLinkingRequest"
                        }
                    ]
                },
                "privacy_level": "anonymous"
            }
        ],
        "public_jwk": {},
        "description": "edX Content Library",
        "custom_fields": {},
        "public_jwk_url": "http://127.0.0.1:18010/api/libraries/v2/lti/1.3/pub/jwk/",
        "target_link_uri": "http://127.0.0.1:18010/api/libraries/v2/lti/1.3/launch/",
        "oidc_initiation_url": "http://127.0.0.1:18010/api/libraries/v2/lti/1.3/login/"
    }
    ```

1. Give the key a name, e.g. "edx tool".
1. On "Redirect URIs" field, add `http://127.0.0.1:18010/api/libraries/v2/lti/1.3/launch/`.
1. Enable the key, switching from `OFF` to `ON`.
1. Copy the "client id" specified here:

   ![Client ID Screenshot](https://user-images.githubusercontent.com/291493/115919353-410f2e00-a468-11eb-9e32-c7a93cde6313.png)

1. On the left menu, go to "Settings"
1. Go to the "Apps" tab.
1. Click on the "+ App" button.
1. On "Configuration Type" select "By Client ID".
1. Enter the client id you copied above.
1. Copy the "deployment ids":

   ![Deployment ID Screenshot](https://raw.githubusercontent.com/dmitry-viskov/repos-assets/master/pylti1p3/examples/canvas-lms/004.png)

### 3. Setup edX.

1. [Generate a key pair for the tool](https://gist.github.com/ygotthilf/baa58da5c3dd1f69fae9).
1. Go to `http://localhost:18010/admin/lti1p3_tool_config/`, login with `edx`.
1. Add a "ti 1.3 tool key".
1. Add the private key and public keys from above.
1. Add a "Lti 1.3 tool"
1. Use the following information:

    ```
        "issuer": "https://canvas.instructure.com",

        // from Canvas: Developer Keys -> value from Details column
        "client_id": "<client id>",

        // static URL
        "auth_login_url": "http://canvas.docker/api/lti/authorize_redirect",

        // static URL
        "auth_token_url": "http://canvas.docker/login/oauth2/token",  

        // static URL to get Platform's public key
        "key_set_url": "http://canvas.docker/api/lti/security/jwks",

        // copy deployment ID from the Canvas created app (screenshot above)
        "deployment_ids": ["<client id>"]
    ```

1. Select the key you created before for "Tool Key".


### 4. Create a library and consume assignment

1. Go to studio and create a library, use the steps described in [`frontend-app-libary-authoring` Devstack Installation](https://github.com/edx/frontend-app-library-authoring/).  Library should be of type "Complex (beta)".
2. Click on "Edit" in one assignment.  In the URL, copy the "block usage key".  For example, if the URL is `http://localhost:3001/library/lib:jvdm:jvdm/blocks/lb:jvdm:jvdm:done:35bb60f2-6d99-4da9-b8cc-9f018671e4f5/` the usage key is `lb:jvdm:jvdm:done:35bb60f2-6d99-4da9-b8cc-9f018671e4f5`.
3. Create the LTI launch URL for this block: For example: `http://127.0.0.1:18010/api/libraries/v2/lti/1.3/launch/?id=lb:jvdm:jvdm:poll:ae5a61a8-f163-44b9-bee0-bddfc219ba33`.
4. Go to Canvas and create an "External Tool" assignment, by performing [this step](https://github.com/dmitry-viskov/pylti1.3/wiki/Configure-Canvas-as-LTI-1.3-Platform#-1) and then [this one](https://github.com/dmitry-viskov/pylti1.3/wiki/Configure-Canvas-as-LTI-1.3-Platform#-2).

### 5. Expected result

The assignment  should be exposed in the Canvas UI

## Deadline

None.

## Other information

This change doesn't depend on changes anywhere else.

This change is *incomplete* and should not be merged to production branches, due to security issues on the new URL endpoints being introduced.

Finally, some blocks are not being rendered correctly on Canvas IFrame in some browsers.  Experiment with different blocks to evaluate the change.

## OCIM Setting

**Settings**
```yaml

EDXAPP_FEATURES:
    ENABLE_LIBRARY_AUTHORING_MICROFRONTEND: true

EDXAPP_LIBRARY_AUTHORING_MICROFRONTEND_URL: "https://library-authoring.{{ EDXAPP_LMS_BASE }}"

TRUSTED_DOMAINS:
  - "library-authoring.{{ EDXAPP_LMS_BASE }}"
  - "blockstore.{{ EDXAPP_LMS_BASE }}"

EDXAPP_ENABLE_CORS_HEADERS: yes
EDXAPP_ENABLE_CROSS_DOMAIN_CSRF_COOKIE: yes
EDXAPP_CORS_ORIGIN_WHITELIST: "{{ TRUSTED_DOMAINS }}"
EDXAPP_LOGIN_REDIRECT_WHITELIST: "{{ TRUSTED_DOMAINS }}"
EDXAPP_CSRF_TRUSTED_ORIGINS: "{{ TRUSTED_DOMAINS }}"

MFE_DEPLOY_STANDALONE_NGINX: yes

MFES:
  - name: library-authoring
    repo: frontend-app-library-authoring
    env_extra:
        STUDIO_BASE_URL: https://{{ EDXAPP_CMS_BASE }}
```

edx-webhook commented:
Thanks for the pull request, @jvdm! I've created [OSPR-5759](https://openedx.atlassian.net/browse/OSPR-5759) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

natabene commented:
@jvdm Thank you for your contribution. Please let me know once it is ready for our review.

arbrandes reviewed (commented):
@jvdm, I have not had time to actually test this, yet, but from a code-read it looks very promising!  Excellent work!  I could not find anything to object to.

I'll test it during the week.

jvdm commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +161,82 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+        if not self.user:
+            uid = uuid.uuid5(uuid.NAMESPACE_URL, self.subject_url)
+            username = f'urn:openedx:content_libraries:username:{uid}'
+            email = f'{uid}@{ContentLibrariesConfig.name}'
```
Are you talking about the username scheme or creating users automatically on `save()`? Or both?

Regarding the username, [one alternative approach is creating randomly generated usernames][1]. Both would work seamlessly, but I believe it is more maintainable to have them namespaced with some scheme, hence the patch as is. The primary benefit is enforcement of LTI 1.3's premise that ISS/SUB claims are unique per LTI integration at the authentication level. It means the username inside Open EdX instances will have a 1:1 mapping with the LTI platform for a given deployment ID.

Drawback is we don't have additional logic to block the creation of users prefixed with `urn: openedx` (or whatever namespace we want to incorporate as reserved). Or protection against the very unlikely event of some instance somewhere already using this prefix.

[1]: https://github.com/edx/edx-platform/blob/master/lms/djangoapps/lti_provider/users.py#L46

marcotuts replied:
Not sure if this is covered/flagged already but we limit PII sharing if a certain LTI waffle flag isn't set or enabled per course currently. I don't know how this would / could extend to this broader blockstore configuration concept, but ideally, we by default aren't using learner PII for any automated IDs unless that sharing is explicitly set/enabled?

jvdm replied:
@marcotuts:

> we by default aren't using learner PII for any automated IDs unless that sharing is explicitly set/enabled?

I don't think that restriction applies here. The LTI launches performed through Content Libraries are detached from any current user in the platform. Except for the `subject_url` I think all the other attributes here are dynamically created. `subject_url` is supposed to be opaque, and even so, we "hash it" under an UUIDv5.

Do you agree? If not, any suggestion on  what would be the next steps here?

bradenmacdonald replied:
> The LTI launches performed through Content Libraries are detached from any current user in the platform.

What if the user is already logged in to Open edX, and then does an LTI launch in an iframe - will it use their existing account? From the code it sounds like yes ("Unless anonymous..."), and I guess that's what I would expect but I'm a little unclear on this point.

> We'll probably need to check with edX that this specific way of auto-creating users works for them. No action needed now, though - just a note for the future.

Yes, we'll need to figure that out now. The concern that I've seen expressed from devops folks in the past is about the users table becoming filled with tens or hundreds thousands of user accounts from individual LTI sessions. Or to be specific, in the past the concern was about different implementations of "anonymous access" to courses generating tons of user account detritus I think, but it's very similar in practice.

That's why with the blockstore runtime, I implemented support for "anonymous users" that doesn't create `User` objects in the database at all, which [was an explicit design goal](https://github.com/edx/edx-platform/tree/master/openedx/core/djangoapps/xblock#high-level-design-goals): 

> To provide first-class support for anonymous usage of XBlocks (use by unregistered users): If the user is unregistered, field data will be saved into the django cache rather than the database.

Question 1: is it possible to use the new runtime's existing approach to "anonymous users" with LTI, or is it incompatible?
Question 2: how does this LTI tool implementation compare with the platform's existing (course/modulestore) LTI tool provider implementation in terms of creating users?
Question 3: If we need to create "real" user entries for LTI users, as done in this PR, can we create a cleanup task that will periodically delete them? Or would that erase important data?

See 

https://github.com/edx/edx-platform/blob/49aaa8621d195a452000942aae8286966a91b96a/openedx/core/djangoapps/xblock/utils.py#L143-L152

and

https://github.com/edx/edx-platform/blob/49aaa8621d195a452000942aae8286966a91b96a/openedx/core/djangoapps/xblock/runtime/ephemeral_field_data.py#L1-L17

jvdm replied:
> > The LTI launches performed through Content Libraries are detached from any
> > current user in the platform.
> 
> What if the user is already logged in to Open edX, and then does an LTI launch
> in an iframe - will it use their existing account? From the code it sounds
> like yes ("Unless anonymous..."), and I guess that's what I would expect but
> I'm a little unclear on this point.

As a matter of fact, no. It won't.  First, because I think the Platform's
cookies are not cross-site (I think they have `SameSite=Lax`, but I would have
to double-check), but more importantly, LTI launches ignore whatever user in the
current session and perform the whole auth/login cycle based on the claims
passed from the platform, in particular, the [user identify claim
`sub`](http://www.imsglobal.org/spec/lti/v1p3/#user-identity-claims).

> > We'll probably need to check with edX that this specific way of
> > auto-creating users works for them. No action needed now, though - just a
> > note for the future.
> 
> Yes, we'll need to figure that out now. The concern that I've seen expressed
> from devops folks in the past is about the users table becoming filled with
> tens or hundreds thousands of user accounts from individual LTI sessions. Or
> to be specific, in the past the concern was about different implementations of
> "anonymous access" to courses generating tons of user account detritus I
> think, but it's very similar in practice.

That is a valid concern.  I pretty much followed the current LTI 1.1
implementation footsteps, and created users attached to custom LTI profiles.
This turned out to be comfortable with the current grading signals that
identify responses per `(user_id, usage_id)`.  By creating one user per LTI
launch, [I was able to pull the proper claims to reply to the LTI Platform to
sync grades](https://github.com/open-craft/edx-platform/blob/35c24c884922f6899955b9b79b3c3983b19e1800/openedx/core/djangoapps/content_libraries/signal_handlers.py#L34).

> That's why with the blockstore runtime, I implemented support for "anonymous
> users" that doesn't create `User` objects in the database at all, which [was
> an explicit design
> goal](https://github.com/edx/edx-platform/tree/master/openedx/core/djangoapps/xblock#high-level-design-goals):
> 
> > To provide first-class support for anonymous usage of XBlocks (use by
> > unregistered users): If the user is unregistered, field data will be saved
> > into the django cache rather than the database.

Oh, that's interesting.  With this approach, how the grading signal handlers
would be called?

> Question 1: is it possible to use the new runtime's existing approach to
> "anonymous users" with LTI, or is it incompatible?

Not sure. If the cached user "entity" is dropped, another launch on the
same `SUB` and `ISS` will re-populate the relationships between user id and the
graded resource? For me to continue to handle grade signals properly, that
would be necessary.  I would need to look at the code and perform
some tests here.

> Question 2: how does this LTI tool implementation compare with the platform's
> existing (course/modulestore) LTI tool provider implementation in terms of
> creating users?

It's pretty much the same approach.  We [clarified that in the
ADR](https://github.com/edx/edx-platform/pull/27089/files).

> Question 3: If we need to create "real" user entries for LTI users, as done in
> this PR, can we create a cleanup task that will periodically delete them? Or
> would that erase important data?

From this patch perspective, no.  I am not sure if the grading subsystem will
have any side-effect if we delete the user associate with some of the grades.

bradenmacdonald replied:
> > support for "anonymous users" that doesn't create User objects in the database at all, which was an explicit design goal
>
> Oh, that's interesting. With this approach, how the grading signal handlers
would be called?

Currently, [they are not](https://github.com/edx/edx-platform/blob/91d5b668a895c89d695227b48afe2c2552c284e2/openedx/core/djangoapps/xblock/runtime/runtime.py#L167-L168) (if the user is anonymous), for that exact reason - I didn't want to pass a fake user object into the existing signal handlers as that could break their API. So perhaps that answers my question: no the ephemeral users (not saved into MySQL) probably won't work for LTI if the implementation is based on the existing Django signal handlers. 

That said, one could very easily modify [this code](https://github.com/edx/edx-platform/blob/91d5b668a895c89d695227b48afe2c2552c284e2/openedx/core/djangoapps/xblock/runtime/runtime.py#L167-L168) to emit a second signal like `ANONYMOUS_USER_SCORE_PUBLISHED` and `ANONYMOUS_PROBLEM_WEIGHTED_SCORE_CHANGED` which the LTI code could listen to, associate with a user, and then send back to the LTI client. If LTI users and grades don't need to persist in the platform for longer than a few hours, I think this would be a great approach, and avoid polluting the `User` table with temporary users. But if you need to be able to retrieve the same LTI user/profile days later, then the approach you have already taken is better/necessary.

> > Question 3: If we need to create "real" user entries for LTI users, as done in
this PR, can we create a cleanup task that will periodically delete them? Or
would that erase important data?
>
> From this patch perspective, no. I am not sure if the grading subsystem will
have any side-effect if we delete the user associate with some of the grades.

I meant as a future addition, not in this PR. And I was assuming it would delete the users and grades, i.e. do a cascading user delete.

For now, can you please just make sure there is some easy way to select all LTI 1.3 users in the database, so that doing a mass delete of them (or a mass delete of "all LTI users older than 1 month", etc.) is at least theoretically possible? It seems what you have here will already permit that, I just want to be explicit.

natabene commented:
@jvdm Is this ready for our review?

jvdm commented:
Yes, @natabene, I believe it is. This reply is enough to trigger the review, or is there some other step missing?

natabene commented:
@jvdm You are all set, I will line this up for a review.

arch-bom-gocd-alerts commented:
📣 💥 Heads-up: You must either **rebase onto master** or **merge master into your branch** to avoid breaking the build.

We recently removed diff-quality and introduced lint-amnesty. This means that the automated quality check that has run on your branch doesn't work the same way it will on master. If you have introduced any quality failures, they might pass on the PR but then break the build on master.

This branch has been detected to not have commit 2e335653 as an ancestor. Here's how to see for yourself:

```
git merge-base --is-ancestor 2e335653 jvdm/content-libraries-as-lti-tool && echo "You're all set" || echo "Please rebase onto master or merge master to your branch"
```

If you have any questions, please reach out to the Architecture team (either #edx-shared-architecture on Open edX Slack or #architecture on edX internal).

jvdm commented:
@marcotuts, I've rebased the branch on top of edx/master. Do you have any additional comments?

arbrandes commented:
@jvdm, can you please add [SE-4514] to the title, to make it easier for us to track it internally?  Thanks!

pomegranited commented:
From conversations on the ticket, it looks like @arbrandes  and @bradenmacdonald  will be reviewing this ticket?
And since @natabene says [TNL are ready to delegate this to core committers](https://openedx.slack.com/archives/C015KC8CN76/p1628534531018500), you should be good to merge this once you're happy with it :)

bradenmacdonald commented:
Great. @natabene you can assign me as a reviewer here if that works?

jvdm commented:
The build has failed. The details are yielding "HTTP ERROR 404 Not Found". I will submit an empty commit to trigger another build in the meantime. Or, if there is a better way to move forward @bradenmacdonald, let me know.

bradenmacdonald commented:
@jvdm Looks like it's resolved now, but in general [there are special commands you can put in a GitHub comment](https://openedx.atlassian.net/wiki/spaces/TE/pages/34013222/Jenkins%2BGuide) that will trigger new builds or re-run failed builds.

bradenmacdonald commented:
> First, grade syncing assumes a 1:1 relationship between XBlocks and grade lineitem. Hence, the logic ignores the potential tree relationship of XBlocks and how the tree should be mapped to a single LTI launch. Lack of a deeper understanding of this relationship within content libraries and the whole XBlock use-cases is the main reason for the current approach. Meaning, I understood users of LTI want to expose individual blocks hosting one assessment. This conclusion was the result of my own tests within Content Libraries.

Perhaps we should dig into this further. What I would suggest is that if there is a [Unit](https://github.com/edx/edx-platform/blob/d6202880ae43845d1ce6f6027f7e10a5f48a034a/common/lib/xmodule/xmodule/unit_block.py#L18) added to a library that contains two child problems, and then if the Unit is used for an LTI launch, any interaction with one of the problems should trigger the overall grade of the Unit to be sent to the LTI consumer, which is calculated as the weighted sum of the grades of the problems.

The existing LTI provider [uses this approach](https://github.com/edx/edx-platform/blob/3152c14192dcccac79b7b7a76eba91b380eba78a/lms/djangoapps/lti_provider/tasks.py#L20-L66) but is quite complex and slow as it must take course grading policies into account. For Blockstore libraries, there are no grading policies to consider and the score of a Unit would be the weighted sum of the scores of its children.

https://github.com/edx/edx-platform/blob/3152c14192dcccac79b7b7a76eba91b380eba78a/lms/djangoapps/lti_provider/tasks.py#L26-L27

bradenmacdonald commented on cms/templates/content_libraries/xblock_iframe.html:
```diff
@@ -0,0 +1,295 @@
+<!DOCTYPE html>
```
There are at least four versions of this iframe JS adapter file that I know of now:
* this
* [`frontend-app-library-authoring/src/library-authoring/edit-block/LibraryBlock/wrap.js`](https://github.com/edx/frontend-app-library-authoring/blob/e189348dba7ea086c3c223055f0054a7bfa2eb8c/src/library-authoring/edit-block/LibraryBlock/wrap.js)
* [Ramshackle's wrap.ts](https://github.com/open-craft/ramshackle/blob/master/ramshackle/js-src/Block/wrap.ts) (admittedly a temporary project)
* LabXchange

We don't need to solve this now, but I've been worried about how fragile this approach is and would love to come up with a better solution or some way to record these JS dependencies in a common place. Or at the very least we can declare this one or the frontend-app-library-authoring one as the "main" version and put comments in the others that they should be based on the "main" version?

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -82,6 +138,25 @@ def library_key(self):
         return LibraryLocatorV2(org=self.org.short_name, slug=self.slug)
 
+    @classmethod
+    def authorize_lti_launch(cls, library_key, *, issuer, client_id=None):
```
What is the purpose of this `, *,` ?

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+    Custom manager of LtiProfile mode.
+    """
+
+    def get_from_claims(self, *, iss, aud, sub):
```
Same here, I don't understand the purpose of the `, *,` especially as you are always calling this method using `kwargs` like `iss=x, aud=y`.

bradenmacdonald replied:
Oh nice, TIL. I'm a fan of that, keyword arguments all the way.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+
+class LtiProfile(models.Model):
+    """
+    Content Librarie LTI's profile for Open edX users.
```
Nit: `Librarie`

bradenmacdonald reviewed (commented):
Adding some comments and questions for now, haven't finished reading through the code yet.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -71,6 +118,15 @@ class ContentLibrary(models.Model):
+        null=True,
+        on_delete=models.SET_NULL,
+        help_text=("Authorize the  LTI tool selected to expose this library's content "
+                   "through LTI launches, leave unselected to disable LTI launches.")
```
Correct my if I'm wrong but isn't this about authorizing an LTI _Platform_ to access this content library? (This content library is the tool). I think the [`LtiTool`](https://github.com/dmitry-viskov/pylti1.3/blob/ec931d3117889b7236048b3426fc9af0321b7bc6/pylti1p3/contrib/django/lti1p3_tool_config/models.py#L37) model is poorly named, as it seems like it holds various data values about a specific _platform_, and is used by a tool to authenticate the platform.

Docstring aside, does this setup allow multiple LTI Platforms (multiple external LMSs) to use a single content library as an LTI tool? From my superficial understanding of this it seems like this will allow multiple content libraries (tools) to be authorized to the same external LMS (platforms), which is good, but that it would not allow a single content library to be used by multiple external LMSs, which definitely needs to be supported as that's one of the main reasons for centralizing content into a shared library.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+        verbose_name=_('open edx user'),
+    )
+
+    platform_id = models.CharField(
```
Can we rename `platform_id` to `issuer` for consistency? I found it a bit confusing at first to figure out what `platform_id` was until I saw that in almost every place it's referenced, it's used like `platform_id=iss`

Or since `platform_id` and `client_id` are both in [`LtiTool`](https://github.com/dmitry-viskov/pylti1.3/blob/ec931d3117889b7236048b3426fc9af0321b7bc6/pylti1p3/contrib/django/lti1p3_tool_config/models.py#L40-L43), would it be helpful to replace these two fields with a foreign key to `LtiTool`? (Or if the issuer/client_id are changed, do we want to ensure that old user records do not get associated with the new ID?)

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+        resource.resource_title = resource_title
+        resource.resource_id = resource_id
+        resource.ags_lineitem = lineitem
+        resource.save()
```
Nit: Django has [`update_or_create`](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#update-or-create) which you can use to code this sort of upsert more declaratively and with fewer lines.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+    resource_id = models.CharField(
+        max_length=255,
+        help_text=_('The platform unique identifier of this resource in the '
+                    'platform, also known as "resource link id".'),
```
The term "platform" here (and in a few other places) is ambiguous to me, because Open edX is frequently refered to as "the platform" both in the codebase and the community. Can we say "the **external** platform" in docstrings like this?

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+        related_name='lti_resources',
+        help_text=_('The authorized LTI profile that launched the resource.'))
+
+    usage_key = models.CharField(
```
Nit: probably better to use [`UsageKeyField`](https://github.com/edx/opaque-keys/blob/46f2886d411e9cf65d25e460be89f8a70e54bf98/opaque_keys/edx/django/models.py#L198)

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+        LtiProfile,
+        on_delete=models.CASCADE,
+        related_name='lti_resources',
+        help_text=_('The authorized LTI profile that launched the resource.'))
```
Nit: Can you just add to the docstring "The authorized LTI profile that launched the resource (**identifies the user**)."

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+
+    def update_score(self, weighted_earned, weighted_possible, timestamp):
+        """
+        Use LTI's score service to update the platform's gradebook.
```
Nit: "the **external** platform" (this Open edX platform has a "gradebook" too)

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+        Use LTI's score service to update the platform's gradebook.
+
+        This method synchronously send a request to the platform to update the
+        assignment score, raising in case the
```
Unfinished sentence ^

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/signal_handlers.py:
```diff
@@ -0,0 +1,43 @@
+        )
+    except LtiGradedResource.DoesNotExist:
+        log.error("LTI 1.3: Score Signal: Unknown resource, ignoring: kwargs=%s",
+                  kwargs)
```
Isn't this handler going to be called for every single problem grade event on the entire platform, including hundreds of thousands of interactions with non-library non-LTI content? If so, we definitely don't want to log anything in the (extremely common) case that there is no `LtiGradedResource` matching the event.

This will also add 1-2 additional database lookups to every single problem submit event, so to make it more performant I think you should at least add a short-circuit line that will return early if the `usage_id` is not a content libraries v2 identifier.

It would also be good to have a feature flag that disables the LTI1.3 library support entirely, including this signal handler - I'm going to mention that as a top level comment.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/views.py:
```diff
@@ -689,3 +743,251 @@ def delete(self, request, usage_key_str, file_path):
+        return {
+            'fragment': fragment,
+            'handler_urls_json': json.dumps(handler_urls),
+            'lms_root_url': settings.LMS_ROOT_URL,
```
Usually this should be `lms_root_url = configuration_helpers.get_value('LMS_ROOT_URL', settings.LMS_ROOT_URL)` because Open edX supports multiple "sites" where the LMS URL can be different on each site, but they share a common database.

```python
from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
```
is the import needed.

bradenmacdonald commented:
@jvdm I've finished reading through the code and have hopefully made all the comments that I'm going to make, so let's focus on resolving them and then I'll do a final test and we can get this merged.

One of my most important comments was about N:N support for external platforms:libraries and I realized that you had mentioned that in the ADR, so I've commented over there too.

My last major piece of feedback (hopefully!) is that some instances will want a way to completely disable LTI 1.3 tool functionality so ideally there should be a setting that will disable it (at least disable the grades signal handler and the REST endpoints, and returning a clear message about how the feature is disabled). The existing provider uses `settings.FEATURES['ENABLE_LTI_PROVIDER']` and you can either use that same feature flag or create a new `settings.FEATURES['ENABLE_LTI_LIBRARY_PROVIDER']` ?

jvdm commented:
> @jvdm I've finished reading through the code and have hopefully made all the comments that I'm going to make, so let's focus on resolving them and then I'll do a final test and we can get this merged.

Thank you for your time and attention to detail.

> One of my most important comments was about N:N support for external platforms:libraries and I realized that you had mentioned that in the ADR, so I've commented over there too.

Sure, let's follow up in the ADR PR.

> My last major piece of feedback (hopefully!) is that some instances will want a way to completely disable LTI 1.3 tool functionality so ideally there should be a setting that will disable it (at least disable the grades signal handler and the REST endpoints, and returning a clear message about how the feature is disabled). The existing provider uses `settings.FEATURES['ENABLE_LTI_PROVIDER']` so you can either use that same feature flag or create a new `settings.FEATURES['ENABLE_LTI_LIBRARY_PROVIDER']` ?

I've addressed the comments, including the feature flag. Its presence disables the Content Libraries LTI views and prevents AGS submission on the grading signal handler. Let me know if you have any other comments.

> Edit: oh and just to ensure it doesn't get lost, there's my earlier comment that we should ensure
> 
> > the score of a Unit would be the weighted sum of the scores of its children.

Right, I have not forgotten it. I am planning to follow up on it in a separate PR. It doesn't seem to block the primary use case at this point.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/tests/test_views_lti.py:
```diff
@@ -16,12 +17,23 @@ class LtiToolJwksViewTest(TestCase):
         Then return empty
         """
+        settings.FEATURES['ENABLE_CONTENT_LIBRARIES'] = True
+        settings.FEATURES['ENABLE_CONTENT_LIBRARIES_LTI_TOOL'] = True
```
Probably better to use [`@override_settings`](https://docs.djangoproject.com/en/3.2/topics/testing/tools/#django.test.override_settings) - search this codebase for examples of using it to change `FEATURES` for a specific test.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/tests/test_views_lti.py:
```diff
@@ -32,6 +44,13 @@ class LibraryBlockLtiUrlViewTest(ContentLibrariesRestApiTest):
+        """
+        Test the LIT URLs are disable when LTI is disabled.
+        """
+        settings.FEATURES['ENABLE_CONTENT_LIBRARIES'] = True
```
Empty test?

jvdm commented:
@bradenmacdonald Resolved comment. And updated the relationship with LTI tool configuration to N:M. AFAIK, this should be ready to merge once checks are completed.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/migrations/0006_auto_20210615_1916.py:
```diff
@@ -0,0 +1,25 @@
+# Generated by Django 2.2.20 on 2021-06-15 19:16
```
This is really minor so don't both fixing it on its own if there's nothing else to fix, but I try to avoid leaving migrations named `auto_....` - it's much more helpful to give them a clear name like `0008_muliple_lti_per_library`

bradenmacdonald reviewed (approved):
:+1: pending a careful test by @arbrandes 

- [x] I have asked @arbrandes to manually test the final version before we merge
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation: docstrings and feature toggles yes, could use more dedicated docs though

arbrandes commented:
@bradenmacdonald,

> +1 pending a careful test by @arbrandes

I have done a comprehensive test of this and related PRs (https://github.com/edx/edx-platform/pull/28464, https://github.com/edx/frontend-app-library-authoring/pull/22, and https://github.com/edx/frontend-app-library-authoring/pull/24), and it all works as it says on the tin.

Care to merge when you're available?  Thanks!

arbrandes commented:
@jvdm, actually, looks like we need a rebase.  Can you take care of it?
